[BUG] multiple consumer working on file when using consume folder as network share

**Describe the bug**
If i share the consume folder to my network and fill in files from a network attached computer i the files were consumed multiple times including errors

**To Reproduce**
Steps to reproduce the behavior:
1. make /consume/ a network shared folder
2. fill in files from network
3. files are consumed multiple times

**Expected behavior**
i was expecting to have just one file workin on

**logs**
```
[2022-03-22 10:51:20,386] [INFO] [paperless.management.consumer] Adding /usr/src/paperless/src/../consume/Test.pdf to the task queue.

[2022-03-22 10:51:20,393] [INFO] [paperless.management.consumer] Adding /usr/src/paperless/src/../consume/Test.pdf to the task queue.

[2022-03-22 10:51:20,573] [INFO] [paperless.management.consumer] Adding /usr/src/paperless/src/../consume/Test.pdf to the task queue.

[2022-03-22 10:51:20,586] [INFO] [paperless.management.consumer] Adding /usr/src/paperless/src/../consume/Test.pdf to the task queue.

[2022-03-22 10:51:20,593] [INFO] [paperless.management.consumer] Adding /usr/src/paperless/src/../consume/Test.pdf to the task queue.

[2022-03-22 10:51:20,598] [INFO] [paperless.consumer] Consuming Test.pdf

[2022-03-22 10:51:20,603] [DEBUG] [paperless.consumer] Detected mime type: application/pdf

[2022-03-22 10:51:20,614] [DEBUG] [paperless.consumer] Parser: RasterisedDocumentParser

[2022-03-22 10:51:20,620] [DEBUG] [paperless.consumer] Parsing Test.pdf...

[2022-03-22 10:51:20,621] [INFO] [paperless.consumer] Consuming Test.pdf

[2022-03-22 10:51:20,625] [DEBUG] [paperless.consumer] Detected mime type: application/pdf

[2022-03-22 10:51:20,637] [DEBUG] [paperless.consumer] Parser: RasterisedDocumentParser

[2022-03-22 10:51:20,642] [DEBUG] [paperless.consumer] Parsing Test.pdf...

[2022-03-22 10:51:22,008] [DEBUG] [paperless.parsing.tesseract] Extracted text from PDF file /usr/src/paperless/src/../consume/Test.pdf

[2022-03-22 10:51:22,029] [DEBUG] [paperless.parsing.tesseract] Extracted text from PDF file /usr/src/paperless/src/../consume/Test.pdf

[2022-03-22 10:51:22,194] [DEBUG] [paperless.parsing.tesseract] Calling OCRmyPDF with args: {'input_file': '/usr/src/paperless/src/../consume/Test.pdf', 'output_file': '/tmp/paperless/paperless-z04o2u17/archive.pdf', 'use_threads': True, 'jobs': 2, 'language': 'eng', 'output_type': 'pdfa', 'progress_bar': False, 'skip_text': True, 'clean': True, 'deskew': True, 'rotate_pages': True, 'rotate_pages_threshold': 12.0, 'sidecar': '/tmp/paperless/paperless-z04o2u17/sidecar.txt'}

[2022-03-22 10:51:22,213] [DEBUG] [paperless.parsing.tesseract] Calling OCRmyPDF with args: {'input_file': '/usr/src/paperless/src/../consume/Test.pdf', 'output_file': '/tmp/paperless/paperless-2a786d6a/archive.pdf', 'use_threads': True, 'jobs': 2, 'language': 'eng', 'output_type': 'pdfa', 'progress_bar': False, 'skip_text': True, 'clean': True, 'deskew': True, 'rotate_pages': True, 'rotate_pages_threshold': 12.0, 'sidecar': '/tmp/paperless/paperless-2a786d6a/sidecar.txt'}

[2022-03-22 10:52:08,511] [DEBUG] [paperless.parsing.tesseract] Using text from sidecar file

[2022-03-22 10:52:08,512] [WARNING] [paperless.parsing.tesseract] Encountered an error while running OCR: No text was found in the original document. Attempting force OCR to get the text.

[2022-03-22 10:52:08,512] [DEBUG] [paperless.parsing.tesseract] Fallback: Calling OCRmyPDF with args: {'input_file': '/usr/src/paperless/src/../consume/Test.pdf', 'output_file': '/tmp/paperless/paperless-2a786d6a/archive-fallback.pdf', 'use_threads': True, 'jobs': 2, 'language': 'eng', 'output_type': 'pdfa', 'progress_bar': False, 'force_ocr': True, 'clean': True, 'deskew': True, 'rotate_pages': True, 'rotate_pages_threshold': 12.0, 'sidecar': '/tmp/paperless/paperless-2a786d6a/sidecar-fallback.txt'}

[2022-03-22 10:52:08,599] [DEBUG] [paperless.parsing.tesseract] Using text from sidecar file

[2022-03-22 10:52:08,600] [WARNING] [paperless.parsing.tesseract] Encountered an error while running OCR: No text was found in the original document. Attempting force OCR to get the text.

[2022-03-22 10:52:08,600] [DEBUG] [paperless.parsing.tesseract] Fallback: Calling OCRmyPDF with args: {'input_file': '/usr/src/paperless/src/../consume/Test.pdf', 'output_file': '/tmp/paperless/paperless-z04o2u17/archive-fallback.pdf', 'use_threads': True, 'jobs': 2, 'language': 'eng', 'output_type': 'pdfa', 'progress_bar': False, 'force_ocr': True, 'clean': True, 'deskew': True, 'rotate_pages': True, 'rotate_pages_threshold': 12.0, 'sidecar': '/tmp/paperless/paperless-z04o2u17/sidecar-fallback.txt'}

[2022-03-22 10:52:54,579] [DEBUG] [paperless.parsing.tesseract] Using text from sidecar file

[2022-03-22 10:52:54,580] [WARNING] [paperless.parsing.tesseract] No text was found in /usr/src/paperless/src/../consume/Test.pdf, the content will be empty.

[2022-03-22 10:52:54,580] [DEBUG] [paperless.consumer] Generating thumbnail for Test.pdf...

[2022-03-22 10:52:54,587] [DEBUG] [paperless.parsing] Execute: convert -density 300 -scale 500x5000> -alpha remove -strip -auto-orient /tmp/paperless/paperless-2a786d6a/archive.pdf[0] /tmp/paperless/paperless-2a786d6a/convert.png

[2022-03-22 10:52:54,746] [DEBUG] [paperless.parsing.tesseract] Using text from sidecar file

[2022-03-22 10:52:54,747] [WARNING] [paperless.parsing.tesseract] No text was found in /usr/src/paperless/src/../consume/Test.pdf, the content will be empty.

[2022-03-22 10:52:54,747] [DEBUG] [paperless.consumer] Generating thumbnail for Test.pdf...

[2022-03-22 10:52:54,753] [DEBUG] [paperless.parsing] Execute: convert -density 300 -scale 500x5000> -alpha remove -strip -auto-orient /tmp/paperless/paperless-z04o2u17/archive.pdf[0] /tmp/paperless/paperless-z04o2u17/convert.png

[2022-03-22 10:52:56,759] [DEBUG] [paperless.parsing.tesseract] Execute: optipng -silent -o5 /tmp/paperless/paperless-2a786d6a/convert.png -out /tmp/paperless/paperless-2a786d6a/thumb_optipng.png

[2022-03-22 10:52:56,925] [DEBUG] [paperless.parsing.tesseract] Execute: optipng -silent -o5 /tmp/paperless/paperless-z04o2u17/convert.png -out /tmp/paperless/paperless-z04o2u17/thumb_optipng.png

[2022-03-22 10:53:07,858] [DEBUG] [paperless.classifier] Document classification model does not exist (yet), not performing automatic matching.

[2022-03-22 10:53:07,864] [DEBUG] [paperless.consumer] Saving record to database

[2022-03-22 10:53:07,968] [DEBUG] [paperless.consumer] Deleting file /usr/src/paperless/src/../consume/Test.pdf

[2022-03-22 10:53:07,992] [DEBUG] [paperless.parsing.tesseract] Deleting directory /tmp/paperless/paperless-2a786d6a

[2022-03-22 10:53:07,993] [INFO] [paperless.consumer] Document 2022-02-23 Test consumption finished

[2022-03-22 10:53:08,037] [DEBUG] [paperless.classifier] Document classification model does not exist (yet), not performing automatic matching.

[2022-03-22 10:53:08,050] [ERROR] [paperless.consumer] The following error occured while consuming Test.pdf: [Errno 2] No such file or directory: '/usr/src/paperless/src/../consume/Test.pdf'

Traceback (most recent call last):

  File "/usr/src/paperless/src/documents/consumer.py", line 283, in try_consume_file

    document = self._store(text=text, date=date, mime_type=mime_type)

  File "/usr/src/paperless/src/documents/consumer.py", line 361, in _store

    stats = os.stat(self.path)

FileNotFoundError: [Errno 2] No such file or directory: '/usr/src/paperless/src/../consume/Test.pdf'

[2022-03-22 10:53:08,051] [DEBUG] [paperless.parsing.tesseract] Deleting directory /tmp/paperless/paperless-z04o2u17

[2022-03-22 10:53:08,635] [ERROR] [paperless.consumer] Cannot consume /usr/src/paperless/src/../consume/Test.pdf: File not found.

[2022-03-22 10:53:09,130] [ERROR] [paperless.consumer] Cannot consume /usr/src/paperless/src/../consume/Test.pdf: File not found.

[2022-03-22 10:53:09,644] [ERROR] [paperless.consumer] Cannot consume /usr/src/paperless/src/../consume/Test.pdf: File not found.
```

**Relevant information**
 - Host OS of the machine running paperless: [e.g. Synology NAS / docker]
 - Browser [Chrome]
 - Version [latest]
 - Installation method: [docker]

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -5,7 +5,7 @@
 repos:
   # General hooks
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v4.1.0
+    rev: v4.2.0
     hooks:
       - id: check-docstring-first
       - id: check-json
@@ -27,7 +27,7 @@ repos:
       - id: check-case-conflict
       - id: detect-private-key
   - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: "v2.6.1"
+    rev: "v2.6.2"
     hooks:
       - id: prettier
         types_or:
@@ -47,7 +47,7 @@ repos:
       - id: yesqa
         exclude: "(migrations)"
   - repo: https://github.com/asottile/add-trailing-comma
-    rev: "v2.2.1"
+    rev: "v2.2.2"
     hooks:
       - id: add-trailing-comma
         exclude: "(migrations)"

--- a/src/documents/management/commands/document_consumer.py
+++ b/src/documents/management/commands/document_consumer.py
@@ -3,7 +3,9 @@ import os
 from pathlib import Path
 from pathlib import PurePath
 from threading import Thread
+from time import monotonic
 from time import sleep
+from typing import Final
 
 from django.conf import settings
 from django.core.management.base import BaseCommand
@@ -53,6 +55,25 @@ def _consume(filepath):
         logger.warning(f"Not consuming file {filepath}: Unknown file extension.")
         return
 
+    # Total wait time: up to 500ms
+    os_error_retry_count: Final[int] = 50
+    os_error_retry_wait: Final[float] = 0.01
+
+    read_try_count = 0
+    file_open_ok = False
+
+    while (read_try_count < os_error_retry_count) and not file_open_ok:
+        try:
+            with open(filepath, "rb"):
+                file_open_ok = True
+        except OSError:
+            read_try_count += 1
+            sleep(os_error_retry_wait)
+
+    if read_try_count >= os_error_retry_count:
+        logger.warning(f"Not consuming file {filepath}: OS reports file as busy still")
+        return
+
     tag_ids = None
     try:
         if settings.CONSUMER_SUBDIRS_AS_TAGS:
@@ -81,19 +102,23 @@ def _consume_wait_unmodified(file):
 
     logger.debug(f"Waiting for file {file} to remain unmodified")
     mtime = -1
+    size = -1
     current_try = 0
     while current_try < settings.CONSUMER_POLLING_RETRY_COUNT:
         try:
-            new_mtime = os.stat(file).st_mtime
+            stat_data = os.stat(file)
+            new_mtime = stat_data.st_mtime
+            new_size = stat_data.st_size
         except FileNotFoundError:
             logger.debug(
                 f"File {file} moved while waiting for it to remain " f"unmodified.",
             )
             return
-        if new_mtime == mtime:
+        if new_mtime == mtime and new_size == size:
             _consume(file)
             return
         mtime = new_mtime
+        size = new_size
         sleep(settings.CONSUMER_POLLING_DELAY)
         current_try += 1
 
@@ -182,14 +207,32 @@ class Command(BaseCommand):
             descriptor = inotify.add_watch(directory, inotify_flags)
 
         try:
+
+            inotify_debounce: Final[float] = 0.5
+            notified_files = {}
+
             while not self.stop_flag:
+
                 for event in inotify.read(timeout=1000):
                     if recursive:
                         path = inotify.get_path(event.wd)
                     else:
                         path = directory
                     filepath = os.path.join(path, event.name)
-                    _consume(filepath)
+                    notified_files[filepath] = monotonic()
+
+                # Check the files against the timeout
+                still_waiting = {}
+                for filepath in notified_files:
+                    # Time of the last inotify event for this file
+                    last_event_time = notified_files[filepath]
+                    if (monotonic() - last_event_time) > inotify_debounce:
+                        _consume(filepath)
+                    else:
+                        still_waiting[filepath] = last_event_time
+                # These files are still waiting to hit the timeout
+                notified_files = still_waiting
+
         except KeyboardInterrupt:
             pass
 

--- a/src/documents/tests/test_management_consumer.py
+++ b/src/documents/tests/test_management_consumer.py
@@ -260,6 +260,21 @@ class TestConsumer(DirectoriesMixin, ConsumerMixin, TransactionTestCase):
                 f'_is_ignored("{file_path}") != {expected_ignored}',
             )
 
+    @mock.patch("documents.management.commands.document_consumer.open")
+    def test_consume_file_busy(self, open_mock):
+
+        # Calling this mock always raises this
+        open_mock.side_effect = OSError
+
+        self.t_start()
+
+        f = os.path.join(self.dirs.consumption_dir, "my_file.pdf")
+        shutil.copy(self.sample_file, f)
+
+        self.wait_for_task_mock_call()
+
+        self.task_mock.assert_not_called()
+
 
 @override_settings(
     CONSUMER_POLLING=1,

